README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# pseudohouseholds

<!-- badges: start -->

[![R-CMD-check](https://github.com/chris31415926535/pseudohouseholds/actions/workflows/R-CMD-check.yaml/badge.svg)](https://github.com/chris31415926535/pseudohouseholds/actions/workflows/R-CMD-check.yaml)
[![Codecov test
coverage](https://codecov.io/gh/chris31415926535/pseudohouseholds/branch/main/graph/badge.svg)](https://app.codecov.io/gh/chris31415926535/pseudohouseholds?branch=main)
[![CRAN
status](https://www.r-pkg.org/badges/version/pseudohouseholds)](https://CRAN.R-project.org/package=pseudohouseholds)
<!-- badges: end -->

A code-based reproducible package for generating representative points
along road networks within regions–“pseudohouseholds”–that can be used
for population-weighted travel analyses.

## Installation

You can install the current published version of pseudohouseholds from
[CRAN](https://cran.r-project.org/) with:

``` r
install.packages("pseudohouseholds")
```

You can install the development version of pseudohouseholds from
[GitHub](https://github.com/) with:

``` r
# install.packages("devtools")
devtools::install_github("chris31415926535/pseudohouseholds")
```

## Summary

This vignette provides an overview of what pseudohouseholds (PHHs) are,
and how you can use the **pseudohouseholds** package to help create
them.

## What are pseudohouseholds?

In brief, PHHs are representative points placed along road segments and
inside regions that provide spatial distributions of some
region-specific property (in practice, usually population). They provide
an approximate way of “spreading out” regional population distributions
that can be helpful when doing travel analyses. We say that PHHs are
“pseudo” households because they do not represent actual buildings or
households, but they approximate the distribution of households by
creating a set of populated points where we think the households are
likely to be. This can allow us to create finer-grained travel or
coverage analyses.

By contrast, many travel analyses assume that a region’s entire
population is concentrated at a single point, usually the region’s
centroid. While this assumption can be valid for smaller regions like
city blocks, it breaks in larger or rural regions where the population
may be more dispersed, or regions with unusual shapes that may not
contain their own centroids.

## The algorithm for creating pseudohouseholds

Let’s walk through the algorithm by giving a demonstration using test
data included with the package. We’ll consider a single dissemination
block (DB) in Ottawa, Ontario, with unique identifier (DBUID)
“35061699003”. DBs are the smallest region at which Statistics Canada
makes population data available, so they’re a good starting point for a
fine-grained analysis.

Here we load our packages, isolate our DB, and print it to the console:

``` r
library(sf)
#> Linking to GEOS 3.10.2, GDAL 3.4.1, PROJ 8.2.1; sf_use_s2() is TRUE
library(ggplot2)
library(ggspatial)
library(pseudohouseholds)

db <- dplyr::filter(pseudohouseholds::ottawa_db_shp, DBUID == "35061699003")

print(db)
#> Simple feature collection with 1 feature and 2 fields
#> Geometry type: MULTIPOLYGON
#> Dimension:     XY
#> Bounding box:  xmin: 369747.8 ymin: 5027016 xmax: 370185.7 ymax: 5027330
#> Projected CRS: NAD83 / MTM zone 9
#> # A tibble: 1 × 3
#>   DBUID       dbpop2021                                                 geometry
#> * <chr>           <dbl>                                       <MULTIPOLYGON [m]>
#> 1 35061699003       125 (((370113.7 5027251, 370075.7 5027241, 370026.3 5027236…
```

Note that its coordinate reference system (CRS) is *projected*.
Specifically, it’s using NAD83/MTM zone 9 (32189) which gives units in
meters. It’s important to use a projected CRS with this package.

Let’s also plot it to get an idea of its scale:

<img src="man/figures/README-plot_db_map-1.png" width="50%" />

This DB is a few hundred meters across, borders Bank St. on its west,
and extends to the east into an irregularly shaped residential area.
From the printout above, we see that it had a population of 125 in 2021.

### If the region is unpopulated, optionally return a default point.

By default, if a region is unpopulated it receives one single point with
population 0 of `phh_type` type 1. This point will be the centroid if it
is inside the region, otherwise it will be an approximation of the
region’s “pole of inaccessibility.” Details are provided below.

### Find road segments intersecting or near the region

First, we identify road segments that either intersect with the region
or come within a specified buffer. The default buffer is 5 meters, which
we adopt here, although larger or smaller buffers may be appropriate in
different settings.

Plotting our DB and the road segments that intersect it:

<img src="man/figures/README-plot_db_roads-1.png" width="50%" />

### Create initial points by sampling along the road segments

Next we sample points along these road segments. The sampling frequency
is given by the parameter `phh_density`, which is passed along directly
to the function `sf::st_line_sample()` and is measured in units of “PHHs
per unit.” The default value is 0.005 which, in meters, translates to
one PHH every 200 meters. Note that this is for our initial sampling,
and that the final PHHs may be closer or farther than this.

<img src="man/figures/README-plot_db_pointsonroads-1.png" width="50%" />

Optionally, users can set a desired minimum number of regional PHHs
using the parameter `min_phhs_per_region`, which can be helpful with
smaller regions (e.g. small city blocks).

### Create candidate PHHs beside the road network

We want our PHHs to be *inside* our DB, so our next step is to perturb
these points by the distance `delta_distance` to try to create
candidates within the region. To do so, we:

- Find the DB’s centroid;
- Create a “pull” set of candidates that are moved towards the centroid
  by `delta_distance`; and,
- Create a “push” set of candidates that are moved away from the
  centroid by `delta_distance`.

<img src="man/figures/README-plot_db_pointsbesideroads-1.png" width="50%" />

As can be seen, most of the blue “pull” candidates are within the DB and
most of the “push” candidates are outside; however, because of the DB’s
odd geometry, one “pull” candidate is outside, and one “push” is inside.
Creating two sets of candidates increases the algorithm’s processing
overhead, but it helps to ensure that our PHHs are distributed the way
we want.

### Keep only candidates within the region

Next, we keep only those candidate PHHs that are inside the region using
a spatial filter.

<img src="man/figures/README-plot_db_pointsinregion-1.png" width="50%" />

### (If necessary) If no valid points are returned, sample radially around our on-street points

The push/pull algorithm may not return valid PHHs in some cases, for
example if the region is concave and has only small intersections with
any road. In this case we use a backup algorithm that samples 16
candidate points distributed radially with radius `delta_distance`
around our on-street point and selects the first candidate point within
the region.

### (If necessary) If *still* no valid points are returned, return a default region point.

If for any reason the algorithm has still not found a valid candidate
PHH, it will now return a single default point. It will return the
centroid if the centroid is inside the region, or else an approximation
of the visual centre of the region. In either case this point will have
`phh_type` 2, indicating no valid PHHs were found using the standard
methods.

### (Optionally) Ensure our PHHs have a minimum population

If we are assigning populations to our PHHs, we next check to ensure
each PHH will have at least some minimum population set by the parameter
`min_phh_pop`. This is helpful for “thinning out” PHHs in large
sparsely-populated rural areas, where the default line sampling density
would lead to a large number of PHHs with populations less than 1.

This can be disabled by setting `min_phh_pop` to 0.

### (Optionally) Remove candidates that are too close to others

Optionally, we can remove any PHHs that are within `min_phh_distance`
units of another. This is helpful to reduce artificial over-crowding
caused by specific road features: for example, if many short road
segments converge, each could receive a candidate PHH from
`sf::st_line_sample()`.

### Distribute regional population among PHHs

We now have a set of points that are near a road, within the region, and
not too close together. The final step is to assign each PHH an equal
share of the region’s population.

## Considerations

### The final number of PHHs may be more or less than you expect.

The algorithm attempts to satisfy several competing criteria–the minimum
number of PHHs per region, the minimum population of those PHHs, and the
minimum distance between them–and it may not be possible to meet them
all. In particular, it will make its best effort to satisfy
`min_phhs_per_region`, but `min_phh_pop` and `min_phh_distance` take
precedence.

### Any input roads may receive PHHs, so ensure only habitable roads are used as inputs.

Standard GIGO principles apply, so if there are any roads where you
don’t want PHHs (e.g. highways, service roads, unpaved trails), take
care to remove them beforehand.

### PHHs do not correspond directly to dwellings, and should only be used as a proxy to “smooth out” populations for analysis.

Any analysis using PHHs should be clear that they rely on approximations
and do not provide “true” population distributions (whatever that might
even mean). While PHHs can provide advantages over centroid-based
analyses, they also have limitations and should be used with these
limitations in mind.

## Details

### The variable `phh_type`

- 1: Normal PHH
- 2: Unpopulated region
- 3: No valid PHHs found

### The default point

If a region is unpopulated or if it is no valid PHHs can be found, the
function will return the region’s “default point.” If the region’s
centroid is inside it, the default point is the centroid. If the
centroid is outside the region, we use a simple algorithm to choose a
point that is approximately in the region’s “pole of inaccessibility,”
or farthest from each border: we randomly sample 100 points inside the
region, and choose the one that is farthest from any border.

Such points are flagged as either `phh_type` 1 (unpopulated) or
`phh_type` 2 (populated but no valid PHHs found), and care should be
taken when using these variables for analysis since they may be farther
from road networks than regular PHHs and may not represent actual
population distributions.


NEWS.md
# pseudohouseholds 0.1.0

* Update examples to fix errors on CRAN fedora platforms

# pseudohouseholds 0.1.0

* Added a `NEWS.md` file to track changes to the package.
* Initial CRAN release.


tests/testthat/test-basic.R


testthat::test_that("multiplication works", {
  testthat::expect_equal(2 * 2, 4)
})


testthat::test_that("Single region function works on synthetic data", {

  testthat::test_that(
    "Full function works ",{
    phhs <- get_phhs_region(region = region_shp, region_idcol = "region_id", region_popcol = "population", roads = road_shp, roads_idcol = "road_id")

    testthat::expect_equal(class(phhs), c("sf", "tbl_df", "tbl", "data.frame"))
    testthat::expect_equal(nrow(phhs), 35)
    testthat::expect_equal(sum(phhs$population), region_shp$population)

    }
   )
  # # works with no population column
  # get_phhs_region(region = input_regions[1,], region_idcol = "DBUID",  roads = ottawa_road_filtered_shp, roads_idcol = "NGD_UID")
  #
  # # works with no road id column
  # get_phhs_region(region = input_regions[1,], region_idcol = "DBUID",  roads = ottawa_road_filtered_shp, roads_idcol = NA)
  #
  #
  # # should fail: different crs
  # get_phhs_region(region = input_regions[1,], region_idcol = "DBUID", region_popcol = "dbpop2021", roads = roads_wgs, roads_idcol = "NGD_UID")
  #
  # # should fail: not sf
  # get_phhs_region(region = dplyr::tibble(), region_idcol = "DBUID", region_popcol = "dbpop2021", roads = roads_wgs, roads_idcol = "NGD_UID")
  # get_phhs_region(region = input_regions[1,], region_idcol = "DBUID", region_popcol = "dbpop2021", roads = dplyr::tibble(), roads_idcol = "NGD_UID")
  #
  # # should fail: no roads or regions
  # get_phhs_region(region = input_regions[1,], region_idcol = "DBUID", region_popcol = "dbpop2021", roads = ottawa_road_filtered_shp[0,], roads_idcol = "NGD_UID")
  # get_phhs_region(region = input_regions[0,], region_idcol = "DBUID", region_popcol = "dbpop2021", roads = ottawa_road_filtered_shp[0,], roads_idcol = "NGD_UID")
  #
  # # should fail: using WGS84
  # get_phhs_region(region = sf::st_transform(input_regions[1,], crs="WGS84"), region_idcol = "DBUID", region_popcol = "dbpop2021", roads = roads_wgs, roads_idcol = "NGD_UID")
  #
  # # should fail: bad column ids
  # get_phhs_region(region = input_regions[1,], region_idcol = "asdf", region_popcol = "dbpop2021", roads = ottawa_road_filtered_shp, roads_idcol = "NGD_UID")
  # get_phhs_region(region = input_regions[1,], region_idcol = "DBUID", region_popcol = "asdf", roads = ottawa_road_filtered_shp, roads_idcol = "NGD_UID")
  # get_phhs_region(region = input_regions[1,], region_idcol = "DBUID", region_popcol = "dbpop2021", roads = ottawa_road_filtered_shp, roads_idcol = "asdsf")


})

# get_phhs_region(region = region_shp, region_idcol = "region_id", region_popcol = "population", roads = road_shp, roads_idcol = "road_id")


tests/testthat/test-filter_phhs.R
# phh_inregion <- dplyr::tibble(id = 1:100)
# use_pops <- TRUE
# region_pop <- 100
# min_phh_pop <- 29
#
# filter_phhs(use_pops, region_pop, phh_inregion, min_phh_pop)
#
# region_pop/nrow(filter_phhs(use_pops, region_pop, phh_inregion, min_phh_pop) )

testthat::test_that("PHH population filtering returns input if population not enabled", {

  phh_inregion <- dplyr::tibble(id = 1:100)
  use_pops <- FALSE
  region_pop <- 10
  min_phh_pop <- 3

  filtered_phhs <- filter_phhs(use_pops, region_pop, phh_inregion, min_phh_pop)

  testthat::expect_equal(filtered_phhs, phh_inregion)
})

testthat::test_that("PHH population filtering returns input if no PHHs need to be removed", {

  # each PHH will get 100/10= 10 population, min is 3, no filtering required
  phh_inregion <- dplyr::tibble(id = 1:10)
  use_pops <- TRUE
  region_pop <- 100
  min_phh_pop <- 3

  filtered_phhs <- filter_phhs(use_pops, region_pop, phh_inregion, min_phh_pop)

  testthat::expect_equal(filtered_phhs, phh_inregion)
})


testthat::test_that("PHH population filtering returns only first PHH if min pop cannot be achieved", {

  # each PHH will get 100/10= 10 population, min is 3, no filtering required
  phh_inregion <- dplyr::tibble(id = 1:10)
  use_pops <- TRUE
  region_pop <- 10
  min_phh_pop <- 30

  filtered_phhs <- filter_phhs(use_pops, region_pop, phh_inregion, min_phh_pop)

  testthat::expect_equal(filtered_phhs, phh_inregion[1,])
})


testthat::test_that("PHH population filtering respects minimum population request", {

  # test set: 200 PHHs, 1000 total population
  phh_inregion <- dplyr::tibble(id = 1:200)
  use_pops <- TRUE
  region_pop <- 1000

  # test for each min pop between 1 and the region pop
  for (min_phh_pop in 1:(region_pop)){
    filtered_phhs <- filter_phhs(use_pops, region_pop, phh_inregion, min_phh_pop)
    phh_pop <- region_pop / nrow(filtered_phhs)

    testthat::expect_gte(phh_pop, min_phh_pop)
  }
})


tests/testthat/test-get_phhs_single.R
# Unit tests for get_phhs_single(), function that finds PHH for a single region

testthat::test_that(
  "Full function works: correct classes, number of rows, and population",{
    phhs <- get_phhs_single(region = region_shp, region_idcol = "region_id", region_popcol = "population", roads = road_shp, roads_idcol = "road_id")

    #Result has correct classes"
    testthat::expect_equal(class(phhs), c("sf", "tbl_df", "tbl", "data.frame"))

    #Result has correct number of rows
    testthat::expect_equal(nrow(phhs), 35)

    # PHH populations sum to original region population
    testthat::expect_equal(sum(phhs$population), region_shp$population)
  }
)

testthat::test_that(
  "Full function works and warns with no population column",{

    phhs <- suppressWarnings(get_phhs_single(region = region_shp, region_idcol = "region_id",  roads = road_shp, roads_idcol = "road_id"))

    #Function gives warning
    testthat::expect_warning(get_phhs_single(region = region_shp, region_idcol = "region_id",  roads = road_shp, roads_idcol = "road_id"))

    # Result has correct classes
    testthat::expect_equal(class(phhs), c("sf", "tbl_df", "tbl", "data.frame"))

    # Result has correct number of rows
    testthat::expect_equal(nrow(phhs), 35)

    # Result does not have a population column
    testthat::expect_true(!"population" %in% colnames(phhs))
  }
)


testthat::test_that(
  "Full function works and warns with no road_id column",{

    phhs <- suppressWarnings(get_phhs_single(region = region_shp, region_idcol = "region_id", region_popcol = "population", roads = road_shp))

    # Function gives warning
    testthat::expect_warning(get_phhs_single(region = region_shp, region_popcol = "population", region_idcol = "region_id",  roads = road_shp))

    #Result has correct classes
    testthat::expect_equal(class(phhs), c("sf", "tbl_df", "tbl", "data.frame"))

    #Result has correct number of rows
    testthat::expect_equal(nrow(phhs), 35)

    # PHH populations sum to original region population
    testthat::expect_equal(sum(phhs$population), region_shp$population)
  }
)


testthat::test_that("Error if input CRSs are different or are WGS84", {
  region_shp_wgs84 <- sf::st_transform(region_shp, crs="WGS84")
  road_shp_wgs84 <- sf::st_transform(road_shp, crs="WGS84")

  testthat::expect_error(get_phhs_single(region = region_shp_wgs84, region_idcol = "region_id", region_popcol = "population", roads = road_shp, roads_idcol = "road_id"))
  testthat::expect_error(get_phhs_single(region = region_shp, region_idcol = "region_id", region_popcol = "population", roads = road_shp_wgs84, roads_idcol = "road_id"))
  testthat::expect_error(get_phhs_single(region = region_shp_wgs84, region_idcol = "region_id", region_popcol = "population", roads = road_shp_wgs84, roads_idcol = "road_id"))
})

testthat::test_that("Error if region or road inputs are not sf objects", {
  testthat::expect_error(get_phhs_single(region = dplyr::tibble(), region_idcol = "region_id", region_popcol = "population", roads = road_shp, roads_idcol = "road_id"))
  testthat::expect_error(get_phhs_single(region = region_shp, region_idcol = "region_id", region_popcol = "population", roads = dplyr::tibble(), roads_idcol = "road_id"))
})

testthat::test_that("Error if no roads or regions supplied (0-row sf objects) or more than one region supplied", {
  regions_shp <- dplyr::bind_rows(region_shp, region_shp)

  testthat::expect_error(get_phhs_single(region = region_shp, region_idcol = "region_id", region_popcol = "population", roads = road_shp[0,], roads_idcol = "road_id"))
  testthat::expect_error(get_phhs_single(region = region_shp, region_idcol = "region_id", region_popcol = "population", roads = road_shp[0,], roads_idcol = "road_id"))
  testthat::expect_error(get_phhs_single(region = region_shp[0,], region_idcol = "region_id", region_popcol = "population", roads = road_shp[0,], roads_idcol = "road_id"))
  testthat::expect_error(get_phhs_single(region = regions_shp, region_idcol = "region_id", region_popcol = "population", roads = road_shp[0,], roads_idcol = "road_id"))
})

testthat::test_that("Error if bad column IDs supplied", {
  testthat::expect_error(get_phhs_single(region = region_shp[1,], region_idcol = "asdf", region_popcol = "population", roads = road_shp, roads_idcol = "road_id"))
  testthat::expect_error(get_phhs_single(region = region_shp[1,], region_idcol = "region_id", region_popcol = "asdf", roads = road_shp, roads_idcol = "road_id"))
  testthat::expect_error(get_phhs_single(region = region_shp[1,], region_idcol = "region_id", region_popcol = "population", roads = road_shp, roads_idcol = "asdsf"))
})




#region = region_shp; region_idcol = "region_id"; region_popcol = "population"; roads = road_shp; roads_idcol = "road_id"; phh_density = 0.005; min_phh_pop = 5; min_phhs_per_region = 1; min_phh_distance = 25; road_buffer_m = 5; delta_distance_m = 5; skip_unpopulated_regions = TRUE



data-raw/DATASET.R
## code to prepare `DATASET` dataset goes here


# TESTING with synthetic data

coords_shp = matrix(data = c(
  0,0,
  0,1,
  0.5,1.5,
  1,1,
  1,0,
  0,0
), ncol = 2, byrow = TRUE) * 1000

coords_extra <- matrix(
  data = c(
    0,0,
    0.5,0.4,
    1,0
  ), ncol = 2, byrow = TRUE) * 1000

coords_extra2 <- matrix(
  data = c(
    0.5, 0.4,
    0.5, 1.5
  ), ncol = 2, byrow = TRUE) * 1000

region_shp <- sf::st_multipoint(coords_shp) |>
  sf::st_cast("POLYGON")|>
  sf::st_sfc() |>
  sf::st_as_sf(crs = 32189) |>
  dplyr::mutate(region_id = "A",
                population = 600)



# region_shp  |> ggplot() + geom_sf()


road1_shp <- sf::st_linestring(coords_shp)
road2_shp <- sf::st_linestring(coords_extra)
road3_shp <- sf::st_linestring(coords_extra2)

road_shp <- sf::st_multilinestring(list(road1_shp, road2_shp, road3_shp)) |>
sf::st_sfc() |>
  sf::st_as_sf(crs = 32189) |>
  dplyr::mutate(road_id = 1:dplyr::n())

#ggplot2::ggplot(road_shp) + ggplot2::geom_sf()

# ggplot2::ggplot() + ggplot2::geom_sf(data = region_shp) + ggplot2::geom_sf(data = road_shp)



# phh_test <- get_phhs_region(region = region_shp, region_idcol = "region_id", region_popcol = NA, roads = road_shp,
# roads_idcol = NA, phh_density = 0.001)

# ggplot2::ggplot() + ggplot2::geom_sf(data = region_shp) + ggplot2::geom_sf(data = road_shp) + ggplot2::geom_sf(data = phh_test, colour = "blue")


# escape unicode in CRS to avoid R CMD CHECK error https://github.com/r-spatial/sf/issues/1341#issuecomment-1120902127
sf::st_crs(region_shp)$wkt <- gsub("°|º", "\\\u00b0", sf::st_crs(region_shp)$wkt)
sf::st_crs(road_shp)$wkt <- gsub("°|º", "\\\u00b0", sf::st_crs(road_shp)$wkt)

usethis::use_data(region_shp, overwrite = TRUE)
usethis::use_data(road_shp, overwrite = TRUE)


######### Ottawa regions, populations, and roads

ottawa_db_shp <- sf::read_sf("~/datascience/data/spatial/ldb_000b21a_e/ldb_000b21a_e.shp") |>
  dplyr::filter(stringr::str_detect(DBUID, "^3506")) |>
  dplyr::select(DBUID)

ottawa_db_pop <- readr::read_csv("~/datascience/data/spatial/geographic_attribute_file/2021_92-151_X.csv") |>
  dplyr::select(DBUID = DBUID_IDIDU, dbpop2021 = DBPOP2021_IDPOP2021) |>
  dplyr::filter(DBUID %in% ottawa_db_shp$DBUID) |>
  dplyr::mutate(DBUID = as.character(DBUID))

ottawa_db_shp <- dplyr::left_join(ottawa_db_shp, ottawa_db_pop, by = "DBUID") |>
  dplyr::select(DBUID, dbpop2021, geometry) |>
  sf::st_transform(crs=32189)

sf::st_crs(ottawa_db_shp)$wkt <- gsub("°|º", "\\\u00b0", sf::st_crs(ottawa_db_shp)$wkt)

usethis::use_data(ottawa_db_shp, overwrite = TRUE)


ottawa_roads_shp <- sf::read_sf("~/datascience/data/spatial/lrnf000r21a_e/lrnf000r21a_e.shp",
                                query = 'SELECT NGD_UID, NAME, RANK, CLASS FROM "lrnf000r21a_e" WHERE CSDNAME_L = \'Ottawa\' OR CSDNAME_R = \'Ottawa\' ')

ottawa_roads_shp <- sf::st_transform(ottawa_roads_shp, crs=32189)

sf::st_crs(ottawa_roads_shp)$wkt <- gsub("°|º", "\\\u00b0", sf::st_crs(ottawa_roads_shp)$wkt)

usethis::use_data(ottawa_roads_shp, overwrite = TRUE)
